[CALCITE-3011] Support for joins with AggregateJoinTransposeRule

vineetgarg02 opened the pull request.

jcamachor commented on core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:
```diff
@@ -3384,6 +3384,174 @@ private void transitiveInference(RelOptRule... extraRules) throws Exception {
+        .build();
+    final String sql = "select e.job\n"
+        + "from (select * from sales.emp where empno = 10) as e\n"
+        + "join sales.dept as d on e.job = d.name\n"
```
This should be outer.

jcamachor commented on core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:
```diff
@@ -3384,6 +3384,174 @@ private void transitiveInference(RelOptRule... extraRules) throws Exception {
+        .build();
+    final String sql = "select d.name\n"
+        + "from (select * from sales.emp where empno = 10) as e\n"
+        + "join sales.dept as d on e.job = d.name\n"
```
This should be outer.

jcamachor commented on core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:
```diff
@@ -3384,6 +3384,174 @@ private void transitiveInference(RelOptRule... extraRules) throws Exception {
+        .build();
+    final String sql = "select e.job, d.name\n"
+        + "from (select * from sales.emp where empno = 10) as e\n"
+        + "join sales.dept as d on e.job = d.name\n"
```
This should be outer.

vineetgarg02 commented:
> Can you add a case for full outer join?

@hsyuan I have added more cases for outer joins

hsyuan commented on core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:
```diff
@@ -3551,6 +3551,251 @@ private void transitiveInference(RelOptRule... extraRules) throws Exception {
+        .build();
+    final String sql = "select e.job\n"
+        + "from (select * from sales.emp where empno = 10) as e\n"
+        + "left outer join sales.dept as d on e.job = d.name\n"
```
left outer?

vineetgarg02 replied:
@hsyuan Sorry I didn't understand what do you mean by `left outer`?

hsyuan commented:
query and comment don't match.





Thanks~
Haisheng Yuan------------------------------------------------------------------
发件人：Vineet G<notifications@github.com>
日　期：2019年04月23日 11:44:36
收件人：apache/calcite<calcite@noreply.github.com>
抄　送：Haisheng Yuan<h.yuan@alibaba-inc.com>; Mention<mention@noreply.github.com>
主　题：Re: [apache/calcite] [CALCITE-3011] Support for joins with AggregateJoinTransposeRule (#1171)

@vineetgarg02 commented on this pull request. 
In core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:
> +        + "full outer join sales.emp as d on e.job = d.job\n"
+        + "group by e.ename,d.mgr";
+    checkPlanning(tester, preProgram, new HepPlanner(program), sql);
+  }
+
+  // full outer join, group by key is susbset of join keys
+  @Test public void testPushAggregateThroughtOuterJoin16() {
+    final HepProgram preProgram = new HepProgramBuilder()
+        .addRuleInstance(AggregateProjectMergeRule.INSTANCE)
+        .build();
+    final HepProgram program = new HepProgramBuilder()
+        .addRuleInstance(AggregateJoinTransposeRule.EXTENDED)
+        .build();
+    final String sql = "select e.job\n"
+        + "from (select * from sales.emp where empno = 10) as e\n"
+        + "left outer join sales.dept as d on e.job = d.name\n"

@hsyuan Sorry I didn't understand what do you mean by left outer?
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

vineetgarg02 commented:
> query and comment don't match. Thanks~ Haisheng Yuan-------------------------------------------
Thanks for noticing this. I have fixed it and updated the pull request.

vineetgarg02 commented:
@jcamachor I have disable the FULL OUTER JOIN as there is an existing issue. I'll enable it once CALCITE-3012 is fixed

jcamachor commented on core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinTransposeRule.java:
```diff
@@ -146,12 +146,22 @@ private static boolean isAggregateSupported(Aggregate aggregate, boolean allowFu
 
+  // OUTER joins are supported for group by without aggregate functions
+  private boolean isJoinSupported(final Join join, final Aggregate aggregate) {
+    return join.getJoinType() != JoinRelType.FULL
```
`&&` has higher precedence than `||`, hence this will not avoid execution of this rule for FULL OUTER joins in all cases. This should be rewritten to:
```
return join.getJoinType() == JoinRelType.INNER ||
   aggregate.getAggCallList().isEmpty() && join.getJoinType() != JoinRelType.FULL;
```
Please update comment above to describe why full outer join is not supported.

vineetgarg02 replied:
> has higher precedence than ||
Parenthesis were left out by mistake. Thanks for noticing this. I'll update the pull request.

jcamachor commented on core/src/main/java/org/apache/calcite/rel/rules/AggregateJoinTransposeRule.java:
```diff
@@ -336,8 +346,9 @@ public void onMatch(RelOptRuleCall call) {
 
     boolean aggConvertedToProjects = false;
-    if (allColumnsInAggregate) {
+    if (allColumnsInAggregate && join.getJoinType() != JoinRelType.FULL) {
```
`&& join.getJoinType() != JoinRelType.FULL) {` is not needed anymore if we bail out correctly above?

vineetgarg02 replied:
@jcamachor Correct, I'll remove this.
